### ISSUE TITLE
FONT legibility in new SCALED mode

### ISSUE THREAD

**Issue by jasonmyers** [pygame/pygame#2314]
I'm trying out the new SCALED mode in pygame 2, ref

https://github.com/pygame/pygame/releases/tag/2.0.0

> a special SCALED mode which takes low pixel games and automatically translates them to a larger window. Imagine your 320x200 resolution game of glorious pixel art... on modern screens that would be a very tiny windows. Now with SCALED it shows up big on a Full HD screen

However since Fonts are also rendered at e.g. 320x200, they are blocky and nearly unreadable when scaled up:

Native resolution 1280 x 800

Game resolution 320x200 using SCALED, with a 6pt font:

<img width="249" alt="320x200 SCALED with 6pt font" src="https://user-images.githubusercontent.com/861996/98550409-13437d80-226a-11eb-9e2e-d231602999c1.png">

Typically I think these games render graphics at the low res, scale them up, and the render the font to blit at the higher res without scaling.  I don't think this approach is possible now using SCALED mode.

The same shot above, except without SCALED mode, and I manually scaled up the 320x200 surface prior to blitting the full-res font:

<img width="247" alt="320x200 manual x4 with 6pt font" src="https://user-images.githubusercontent.com/861996/98550438-1b032200-226a-11eb-88f3-6abca2d9d25f.png">

**Comment by ankith26** [comment]
Thanks for the bug report.
Can you please provide the code used to generate both cases, so that it makes it easier for us to understand the issue

**Comment by jasonmyers** [comment]
Sure, I modified the code in https://www.pygame.org/docs/tut/PygameIntro.html to reproduce both cases above:

Using new SCALED mode:
```
import sys, pygame
pygame.init()

size = width, height = 320, 200
black = 0, 0, 0

screen = pygame.display.set_mode(size, flags=pygame.SCALED)

# Create a ball sprite
ball = pygame.surface.Surface((16, 16))
pygame.draw.circle(ball, pygame.color.Color('red'), (8, 8), 8, width=1)
ballrect = ball.get_rect().move((16, 16))

font = pygame.font.Font(None, 6)

while 1:
    for event in pygame.event.get():
        if event.type == pygame.QUIT: sys.exit()

    screen.fill(black)
    screen.blit(ball, ballrect)

    text = font.render('Example', False, pygame.color.Color('white'))

    screen.blit(text, (0, 0))

    pygame.display.flip()
```

Using old mode and manually scaling up the sprites (but not the font):
```
import sys, pygame
pygame.init()

size = width, height = 320 * 4, 200 * 4  # 1280 x 800
black = 0, 0, 0

screen = pygame.display.set_mode(size)
lowres = pygame.surface.Surface((320, 200), flags=pygame.SRCALPHA)  # For pixel graphics
hires = pygame.surface.Surface(size, flags=pygame.SRCALPHA)  # For high-res font

# Create a ball sprite
ball = pygame.surface.Surface((16, 16))
pygame.draw.circle(ball, pygame.color.Color('red'), (8, 8), 8, width=1)
ballrect = ball.get_rect().move((16, 16))

font = pygame.font.Font(None, 6 * 4)

while 1:
    for event in pygame.event.get():
        if event.type == pygame.QUIT: sys.exit()

    screen.fill(black)

    lowres.blit(ball, ballrect)

    text = font.render('Example', False, pygame.color.Color('white'))
    hires.blit(text, (0, 0))

    # Scale up low-res to match hi-res screen
    scaled = pygame.transform.scale(lowres, size)

    screen.blit(scaled, (0, 0))
    screen.blit(hires, (0, 0))
    pygame.display.flip()
```

**Comment by jasonmyers** [comment]
And here's new images for the above two programs:

<img width="134" alt="Screen Shot 2020-11-09 at 11 50 22 AM" src="https://user-images.githubusercontent.com/861996/98570923-f61aa900-2281-11eb-80a0-3307a4826f88.png">
<img width="129" alt="Screen Shot 2020-11-09 at 11 50 33 AM" src="https://user-images.githubusercontent.com/861996/98570924-f6b33f80-2281-11eb-86f9-adbd7688d6fb.png">

**Comment by MyreMylar** [comment]
Is 6 pixel text normally legible anyway?

**Comment by illume** [comment]
Thanks @jasonmyers for the write up of this use case.

I've also thought about this problem with 'SCALED' and device independent resolution rendering.
Whenever drawing text or a HUD it would be useful to be able to make the text more readable, or show more details over the pixely stuff.

**Comment by robertpfeiffer** [comment]
I understand the feature request, and it wouldn't even be *that* hard to implement something like this, but it is much harder to get the API and specification of the feature right. Ideally, SDL2 would have signed distance fields for font rendering available, but they don't have that.

Rendering font in a "crisp" way means rendering exactly at the native resolution of the screen, breaking the abstraction of scaling. So either we implement a new font API (not pygame.font or pygame.ftfont), and rendering method, like the high-res bitmap texture atlas in libgdx, or crisp pixel-art bitmap fonts, or we get SDL2 renderer to support signed distance fields and add some helper function to generate them with pygame.font.

Barring that, you can always implement your own shader-based multi-layer system with OpenGL, or make something based on SDL2 renderer, and render text at a *higher* resolution to guarantee sharpness.

**Personally, I just use a pixel-art font that renders properly at low resolutions. It's the simplest thing that works**
As a stopgap, we could add one or more hinted low-res font to PyGame and use them when a font size of 10 and lower is requested, but you might consider that an ugly hack.

Wow that was a wall of text.
**TL;DR** You *really* can't expect that particular font to be readable at 6px. You can't expect pygame.font to behave differently depending on display resolution. Whatever solution we adopt will have a different API to what we have now.

**Comment by robertpfeiffer** [comment]
@jasonmyers Can you comment on what *you want* to happen? There are multiple ways of doing what you want, I'd personally go with OpenGL rendering game world and HUD layers to separate textures, or doing the same with pygame._sdl2 renderer. It's obviously impossible to just have font rendering just rendering at a different resolution and then draw the result via something else than blitting the pixels based on some transparent do-what-I-mean mechanism that reproduces the high-res screenshot you posted.

So I'd assume you'd be happy enough with pygame.font.Font(None, 6) returning a more appropriate font?

**Comment by jasonmyers** [comment]
A more legible default font at small-sizes sounds like it might be a good enough solution, yes

**Comment by robertpfeiffer** [comment]
Since @Starbuck5 has touched the SysFont code recently: Would it be a simple change to add a switch that selects a suitable **hinted or pixellated** font at small sizes? Should we emit a warning when the "freesansbold.ttf" fallback is used at a very small font size? Can we just close this otherwise?

**Comment by robertpfeiffer** [comment]
Oh shoot I forgot we are uing SDL_TTF by default. Do we even enable proper hinting? I found no reference to this in the code: https://www.libsdl.org/projects/SDL_ttf/docs/SDL_ttf_26.html 

ftfont may or may not do the right thing, there is certainly a lot of hinting stuf in there.

**Comment by Starbuck5** [comment]
> SDL2 would have signed distance fields for font rendering available, but they don't have that.

Coming in the soon to be released SDL_ttf 2.0.18

**Comment by robertpfeiffer** [comment]
> Coming in the soon to be released SDL_ttf 2.0.18

Wait what? Rendering to SDF buffer or rendering SDF to screen via SDL_Renderer?

**Comment by Starbuck5** [comment]
@robertpfeiffer 

Here's the header:
https://github.com/libsdl-org/SDL_ttf/blob/27d8fb759920cd399dc4dcf48c0006f446766655/SDL_ttf.h#L371-L373

You can also check the implementation in SDL_ttf.c if you're curious

I don't know much about it.
